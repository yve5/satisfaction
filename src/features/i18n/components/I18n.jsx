import React from 'react';
import { connect } from 'react-redux';
import { formatDate, tsl } from '../resources/Utilities';
// import '../resources/style.scss';

class I18n extends React.Component {
  constructor() {
    super();
    this.format = this.format.bind(this);
  }

  format(value) {
    const {
      format,
      root: {
        i18n: { lang },
      },
    } = this.props;

    return formatDate(value, format, lang);
  }

  render() {
    const {
      args,
      attrs,
      children,
      date,
      root: {
        i18n: { dictionary, lang },
      },
      title,
    } = this.props;

    if ('date' in this.props) {
      return this.format(date);
    }

    if (!children) {
      return '';
    }

    const argsCopy = args || {};

    Object.keys(argsCopy).forEach((key) => {
      const property = argsCopy[key];

      if (typeof property === 'object' && property !== null && 'date' in property) {
        property.lang = lang;
      }
    });

    if (typeof children === 'string') {
      return tsl(dictionary, children, argsCopy);
    }

    if (typeof children.length !== 'undefined') {
      throw new Error('I18n can not translate several html tags in a row');
    }

    if (children.type === 'option') {
      const { value, children: label } = children.props;
      return <option value={value}>{tsl(dictionary, label, argsCopy)}</option>;
    }

    const attrsCopy = attrs || [];
    if (title) {
      attrsCopy.push('title');
    }

    if (!Array.isArray(attrsCopy) || (Array.isArray(attrsCopy) && !attrsCopy.length)) {
      throw new Error('I18n attributes are missing');
    }

    const translated = {};
    attrsCopy.forEach((element) => {
      if (typeof children.props[element] !== 'undefined') {
        translated[element] = tsl(dictionary, children.props[element], argsCopy);
      }
    });

    return <children.type {...children.props} {...translated} />;
  }
}

const mapStateToProps = (state) => ({
  root: state,
});

export default connect(mapStateToProps)(I18n);
