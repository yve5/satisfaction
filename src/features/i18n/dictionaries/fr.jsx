export const I18N_DICTIONARY = {};
