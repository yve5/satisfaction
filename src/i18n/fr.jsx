const I18N_DICTIONARY = {
  LANG: 'fr',
  'Are you happy with your salary?': 'Votre salaire vous convient-il ?',
  'Are you satisfied with the interaction with people inside or outside the company?':
    'L’intéraction avec les personnes en ou hors de l’entreprise vous satisfait-elle ?',
  'Confidence factor': 'Facteur de confiance',
  'Confidence factor calculation': 'Calcul du facteur de confiance',
  'Do you actually enjoy doing the tasks of your job?':
    'Aimez-vous réellement faire les tâches de votre poste ?',
  'Do you appreciate the quality of the teamwork?':
    'Appréciez-vous la qualité du travail d’équipe ?',
  'Do you have time for your personal projects?':
    'Avez-vous du temps pour vos projets personnels ?',
  'Do you like your workplace?': 'Aimez-vous votre lieu de travail ?',
  'Do your career prospects suit you?':
    'Vos perspectives d’évolution professionnelles vous convient-elles ?',
  'Do your colleagues agree with you?':
    'Vos collègues sont-ils d’accord avec vous ?',
  'Does doing this work bring you closer to your career goals?':
    'Exercer ce travail vous rapproche-t-il de vos objectifs professionnels ?',
  'Is your company doing well?': 'Votre société se porte-t-elle bien ?',
  lang: 'fr',
  languageId: 'Français',
  Print: 'Impression',
  Result: 'Résultat',
  'Source : Brett Nelson': 'Source : Brett Nelson',
  'The job is for you.': 'Le job vous correspond.',
  'Think twice about taking the job, or leaving it right away.':
    'Songez à deux fois avant d’accepter le job, ou le quitter sur-le-champ.',
  'You got the perfect job!': 'Vous avez le job parfait !',
  "You're on the right track to finding the perfect job.":
    'Vous êtes sur la bonne voie pour trouver le job parfait.',
};

export default I18N_DICTIONARY;
