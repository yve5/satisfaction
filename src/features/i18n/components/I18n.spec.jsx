import React from 'react';
import configureMockStore from 'redux-mock-store';
import I18n from './I18n';

import { create } from 'react-test-renderer';
import { Provider } from 'react-redux';
import { I18N_DICTIONARY } from '../dictionaries/fr';

const mockStore = configureMockStore();

const i18n = {
  collection: [],
  dictionary: {
    'John got $pineapples pineapples, $oranges bananas and $oranges oranges.':
      'John a reçu $pineapples ananas, $oranges bananes et $oranges oranges.',
    'Orange is a French telecommunications company.':
      'Orange est une entreprise française de télécommunications.',
    'Pending mode': 'Mode décision',
    'Update of $update, created at $start by $guest':
      'Mise à jour le $update, créé le $start par $guest',
  },
  lang: 'fr',
};

const store = mockStore({ i18n });

beforeEach(() => {
  console.error = jest.fn();
});

describe('I18n', () => {
  it('should match expected snapshot without props', () => {
    const component = create(
      <Provider store={store}>
        <I18n />
      </Provider>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should match expected snapshot with a string to translate', () => {
    const component = create(
      <Provider store={store}>
        <I18n>Pending mode</I18n>
      </Provider>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should match expected snapshot with an option tag to translate', () => {
    const component = create(
      <Provider store={store}>
        <I18n>
          <option value="42">Pending mode</option>
        </I18n>
      </Provider>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should handle an error with 2 chilren', () => {
    const error = () => {
      create(
        <Provider store={store}>
          <I18n>
            <option value="42">Pending mode</option>
            <option value="42">Pending mode</option>
          </I18n>
        </Provider>
      );
    };
    expect(error).toThrow('I18n can not translate several html tags in a row');
  });

  it('should handle an error with wrong attribute type', () => {
    const error = () => {
      create(
        <Provider store={store}>
          <I18n attrs="Hello World">
            <img alt="Pending mode" src="picture" title="Pending mode" />
          </I18n>
        </Provider>
      );
    };
    expect(error).toThrow('I18n attributes are missing');
  });

  it('should handle an error with empty attribute array', () => {
    const error = () => {
      create(
        <Provider store={store}>
          <I18n attrs={[]}>
            <img alt="Pending mode" src="picture" title="Pending mode" />
          </I18n>
        </Provider>
      );
    };
    expect(error).toThrow('I18n attributes are missing');
  });

  it('should match expected snapshot with attributes', () => {
    const component = create(
      <Provider store={store}>
        <I18n attrs={['alt', 'title', 'hello']}>
          <img alt="Pending mode" src="picture" title="Pending mode" />
        </I18n>
      </Provider>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should match expected snapshot with title attribute shortcut', () => {
    const component = create(
      <Provider store={store}>
        <I18n title>
          <img alt="Pending mode" src="picture" title="Pending mode" />
        </I18n>
      </Provider>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should match expected snapshot with an option tag to translate and args', () => {
    const component = create(
      <Provider store={store}>
        <I18n args={{ $guest: 'John' }}>
          <option value="42">
            Update of $update, created at $start by $guest
          </option>
        </I18n>
      </Provider>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});

describe('I18n date parser', () => {
  const dateExample = new Date(2020, 2, 14, 13, 37, 14);

  it('should match expected snapshot with date', () => {
    const component = create(
      <Provider store={store}>
        <I18n date={dateExample} />
      </Provider>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should match expected snapshot with undefined date', () => {
    const component = create(
      <Provider store={store}>
        <I18n date={undefined} />
      </Provider>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should match expected snapshot with date and format', () => {
    const component = create(
      <Provider store={store}>
        <I18n date={dateExample} format="PPp" />
      </Provider>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should match expected snapshot with string, date and format', () => {
    const component = create(
      <Provider store={store}>
        <I18n date={dateExample}>Pending mode</I18n>
      </Provider>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should match expected snapshot with string, date and format', () => {
    const component = create(
      <Provider store={store}>
        <I18n date={dateExample}>Pending mode</I18n>
      </Provider>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  const startExample = new Date(2017, 4, 14, 17, 51, 31);
  const updateExample = new Date(2020, 2, 14, 13, 37, 14);

  it('should match expected snapshot with date args', () => {
    const component = create(
      <Provider store={store}>
        <I18n
          args={{
            $guest: 'Fitzgerald',
            $start: { date: startExample },
            $update: { date: updateExample },
          }}
        >
          Update of $update, created at $start by $guest
        </I18n>
      </Provider>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should match expected snapshot with date args', () => {
    const component = create(
      <Provider store={store}>
        <I18n
          args={{
            $guest: 'Fitzgerald',
            $start: { date: startExample },
            $update: { date: updateExample },
          }}
        >
          Update of $update, created at $start by $guest
        </I18n>
      </Provider>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should match expected snapshot with attributes and date args', () => {
    const component = create(
      <Provider store={store}>
        <I18n
          args={{
            $guest: 'Fitzgerald',
            $start: { date: startExample },
            $update: { date: updateExample },
          }}
          attrs={['alt', 'title', 'hello']}
        >
          <img
            alt="Pending mode"
            src="picture"
            title="Update of $update, created at $start by $guest"
          />
        </I18n>
      </Provider>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should match expected snapshot with input tag', () => {
    const component = create(
      <Provider store={store}>
        <I18n
          args={{ $dateToUpdate: { date: startExample } }}
          attrs={['value']}
        >
          <input
            className="form-control form-control-sm"
            disabled
            type="text"
            value="$dateToUpdate"
          />
        </I18n>
      </Provider>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
