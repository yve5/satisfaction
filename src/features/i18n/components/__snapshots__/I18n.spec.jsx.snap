// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`I18n date parser should match expected snapshot with attributes and date args 1`] = `
<img
  alt="Mode décision"
  src="picture"
  title="Mise à jour le 14/03/2020, créé le 14/05/2017 par Fitzgerald"
/>
`;

exports[`I18n date parser should match expected snapshot with date 1`] = `"14/03/2020"`;

exports[`I18n date parser should match expected snapshot with date and format 1`] = `"14 mars 2020, 13:37"`;

exports[`I18n date parser should match expected snapshot with date args 1`] = `"Mise à jour le 14/03/2020, créé le 14/05/2017 par Fitzgerald"`;

exports[`I18n date parser should match expected snapshot with date args 2`] = `"Mise à jour le 14/03/2020, créé le 14/05/2017 par Fitzgerald"`;

exports[`I18n date parser should match expected snapshot with input tag 1`] = `
<input
  className="form-control form-control-sm"
  disabled={true}
  type="text"
  value="14/05/2017"
/>
`;

exports[`I18n date parser should match expected snapshot with string, date and format 1`] = `"14/03/2020"`;

exports[`I18n date parser should match expected snapshot with string, date and format 2`] = `"14/03/2020"`;

exports[`I18n date parser should match expected snapshot with undefined date 1`] = `"-"`;

exports[`I18n should match expected snapshot with a string to translate 1`] = `"Mode décision"`;

exports[`I18n should match expected snapshot with an option tag to translate 1`] = `
<option
  value="42"
>
  Mode décision
</option>
`;

exports[`I18n should match expected snapshot with an option tag to translate and args 1`] = `
<option
  value="42"
>
  Mise à jour le $update, créé le $start par John
</option>
`;

exports[`I18n should match expected snapshot with attributes 1`] = `
<img
  alt="Mode décision"
  src="picture"
  title="Mode décision"
/>
`;

exports[`I18n should match expected snapshot with title attribute shortcut 1`] = `
<img
  alt="Pending mode"
  src="picture"
  title="Mode décision"
/>
`;

exports[`I18n should match expected snapshot without props 1`] = `""`;
